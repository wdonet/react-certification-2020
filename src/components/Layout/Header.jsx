import React, { useContext, useRef } from 'react';
import styled, { ThemeContext } from 'styled-components';
import { TextField, IconWrapper, Avatar, Switch } from '../ui/index';
import hamburger from '../../assets/icons/hamburguer.png';
import defaultUser from '../../assets/icons/default_user.jpg';
import SearchContext from '../../providers/SearchContext';
import AppContext from '../../providers/AppContext';

const StyledHeader = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: 64px;
  background-color: #00695c;
  overflow: hidden;
  position: fixed;
  top: 0;
`;

const StyledSection = styled.div`
  display: flex;
  align-items: center;
`;

const Header = () => {
  const { setHomeVideosView } = useContext(AppContext);
  const { switchTheme } = useContext(ThemeContext);
  const { search } = useContext(SearchContext);
  const ref = useRef(null);

  const setHomeVideosViewAndSearch = (query) => {
    setHomeVideosView();
    search(query);
  };

  return (
    <StyledHeader role="toolbar" data-testid="header">
      <StyledSection>
        <div>
          <IconWrapper role="button" src={hamburger} alt="hamburguer" />
        </div>
        <div data-testid="search-input">
          <TextField
            role="search"
            ref={ref}
            onKeyPress={({ charCode }) =>
              charCode === 13 && setHomeVideosViewAndSearch(ref.current.value)
            }
          />
        </div>
      </StyledSection>
      <StyledSection>
        <div data-testid="theme-mode-switch">
          <Switch onClick={switchTheme} />
        </div>
        <div data-testid="user-avatar">
          <Avatar role="button" alt="profile" src={defaultUser} />
        </div>
      </StyledSection>
    </StyledHeader>
  );
};

export default Header;
