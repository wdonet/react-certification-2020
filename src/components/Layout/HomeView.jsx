import React from 'react';
import styled from 'styled-components';

const StyledHomeView = styled.div`
  padding-top: 64px;
  height: calc(100% - 64px);
`;

const HomeView = ({ children }) => {
  return <StyledHomeView role="main" data-testid="main-content">{children}</StyledHomeView>;
};

export default HomeView;
