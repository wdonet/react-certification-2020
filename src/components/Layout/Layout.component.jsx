import React from 'react';

function Layout({ children }) {
  return <main className="container">{children}</main>;
}

export default Layout;
