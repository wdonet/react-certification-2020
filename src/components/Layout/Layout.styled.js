import styled from 'styled-components';

const MainContainer = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
`;

const Content = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: space-around;
  align-items: stretch;
`;

export { MainContainer, Content };
