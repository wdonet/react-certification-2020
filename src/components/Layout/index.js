export { default as LayoutWrapper } from './LayoutWrapper';
export { default as Header } from './Header';
export { default as HomeView } from './HomeView';
