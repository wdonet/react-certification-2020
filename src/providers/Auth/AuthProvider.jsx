import React, { useState, useEffect, useCallback } from 'react';

import { AUTH_STORAGE_KEY } from '../../utils/constants';
import { loginUser } from '../../mocks/loginApi';
import { storage } from '../../utils/storage';

const AuthContext = React.createContext(null);

function AuthProvider({ authenticatedJest, children }) {
  const [authenticated, setAuthenticated] = useState(authenticatedJest);

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY);
    const isAuthenticated = Boolean(lastAuthState);

    setAuthenticated(isAuthenticated || authenticatedJest);
  }, []);

  const login = async ({ username, password }) => {
    try {
      const authorized = await loginUser({ username, password });
      if (authorized) {
        setAuthenticated(true);
        storage.set(AUTH_STORAGE_KEY, true);
      }
      return { success: true };
    } catch (error) {
      return { error };
    }
  };

  const logout = useCallback(() => {
    setAuthenticated(false);
    storage.set(AUTH_STORAGE_KEY, false);
  }, []);

  return (
    <AuthContext.Provider value={{ login, logout, authenticated }}>
      {children}
    </AuthContext.Provider>
  );
}

export { AuthContext };
export default AuthProvider;
