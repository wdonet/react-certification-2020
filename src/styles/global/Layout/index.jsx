import { createGlobalStyle } from 'styled-components'

const LayoutStyle = createGlobalStyle`
  :root {
    --header-max-height: 4rem;
    --footer-min-height: 4rem;
  }

  @media (min-width: 768px) {
    :root {
      --footer-min-height: 4rem;
    }
  }

  header {
    max-height: var(--header-max-height);
  }

  footer {
    height: var(--footer-min-height);
  }

  main {
    padding-bottom: var(--footer-min-height);
  }
`

export default LayoutStyle
