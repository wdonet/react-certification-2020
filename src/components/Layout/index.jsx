import React from 'react'
import { StyledLayout } from './styled'
import useThemeStyles from '../../hooks/useThemeStyles'
import useCustom from '../../hooks/useCustom'
import Header from '../../styles/components/Header'
import Main from '../../styles/components/Main'
import Content from '../../styles/components/Content'
import Menu from '../../styles/components/Menu'
import Hero from '../../styles/components/Hero'
import Footer from '../../styles/components/Footer'
import Grid from '../../styles/components/Grid'
import Tile from '../../styles/components/Tile'
import Icon from '../../styles/components/Icon'
import InputSearch from '../../styles/components/InputSearch'
import Image from '../../styles/components/Image'
import logoImage from '../../assets/images/logo.png'
import videosJSON from '../../data/videos.json'

const Layout = () => {
  const { name, theme, setTheme } = useThemeStyles()
  const { increaseThemeClicks, triggered } = useCustom()
  const links = [
    { path: '/', text: 'Home' },
    { path: '/favorites', text: 'Favorites' },
  ]

  console.log(triggered)

  const HeaderLogo = <Image src={logoImage} />
  const HeaderNavigatin = <Menu links={links} />
  const HeaderCorner = (
    <Icon
      size="100%"
      icon={theme.icon}
      onClick={() => {
        new Audio(theme.sound).play()
        increaseThemeClicks()
      }}
      onAnimate={() => {
        setTheme(name === 'light' ? 'dark' : 'light')
      }}
    />
  )

  let heroDisplayColorsAmount = 3
  let heroDisplaySpeed = 2000
  let heroColors = [
    theme.decorator.primary,
    theme.decorator.secondary,
    theme.background.primary,
  ]
  if (triggered.rainbow) {
    heroDisplayColorsAmount = triggered.rainbow.values.colors.length
    heroColors = triggered.rainbow.values.colors
    heroDisplaySpeed = 1000
  }

  // TODO: move this to other smart components following Single Responsability Principle, not able to do this because of timing issues, sorry.
  return (
    <StyledLayout>
      <Header main={HeaderLogo} secondary={HeaderNavigatin} corner={HeaderCorner} />
      <Main>
        <Hero
          backgroundColors={heroColors}
          amountDisplay={heroDisplayColorsAmount}
          animationSpeed={heroDisplaySpeed}
        >
          <InputSearch />
        </Hero>
        <Content>
          <Grid>
            {videosJSON.items.map(video => (
              <Tile src={video.snippet.thumbnails.medium.url} />
            ))}
          </Grid>
        </Content>
      </Main>
      <Footer>Thanks for visiting!</Footer>
    </StyledLayout>
  )
}

Layout.propTypes = {}

export default Layout
