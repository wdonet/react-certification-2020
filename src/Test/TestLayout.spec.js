import React from 'react';
import { render } from '@testing-library/react';
import AuthProvider from '../providers/Auth/Auth.provider';
import Layout from '../components/Layout';

describe('Layout Components', () => {
  it('should take a snapshot', () => {
    const { asFragment } = render(
      <AuthProvider>
        <Layout />
      </AuthProvider>
    );
    expect(
      asFragment(
        <AuthProvider>
          <Layout />
        </AuthProvider>
      )
    ).toMatchSnapshot();
  });
});
