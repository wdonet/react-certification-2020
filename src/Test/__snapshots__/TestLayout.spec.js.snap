// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Components should take a snapshot 1`] = `
<DocumentFragment>
  <main
    class="container"
  />
</DocumentFragment>
`;
